### PR TITLE
Allow for explicit stop of WorkerServer and DispatchServer

### PR DESCRIPTION
**Pull request by EnricoMi**
There is currently no documented way to stop a `WorkerServer` or `DispatchServer`. It gets automatically stopped on GC, but an explicit `stop` method provides better control on when the shutdown occurs.

This adds `stop` to `WorkerServer` or `DispatchServer`, as well as `__exit__`, both calling into existing `_stop`.

### CODE REVIEW

File: tensorflow/python/data/experimental/service/server_lib.py [hunk 1]
```diff
@@ -227,6 +248,9 @@ def _stop(self):
   def __del__(self):
     self._stop()
 
+  def __exit__(self, exc_type, exc_val, exc_tb):
```
Comment (yangustc07) [inline tensorflow/python/data/experimental/service/server_lib.py:251]: Let's follow the convention of https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/distribute/coordinator/cluster_coordinator.py.

We have `stop` and `join`, but `__exit__` is not needed.

**Review by yangustc07 — CHANGES_REQUESTED** [review]
Thanks for sending the change! Let's follow the convention of https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/distribute/coordinator/cluster_coordinator.py and define `stop` and `join`. I think we don't need `__exit__` for now since it doesn't add much value now that we have `stop`.

**Comment by gbaned** [comment]
Hi @EnricoMi Can you please check @yangustc07's comments and keep us posted ? Thank you!

**Comment by EnricoMi** [comment]
@gbaned @yangustc07 I will do the requested changes

**Comment by EnricoMi** [comment]
I have removed `__exit__` from dispatcher and worker.

**Review by yangustc07 — APPROVED** [review]
Thank you for the change!

File: tensorflow/python/data/experimental/service/server_lib.py [hunk 2]
```diff
@@ -200,6 +203,15 @@ def join(self):
     """
     self._server.join()
 
+  def stop(self):
```
Comment (yangustc07) [inline tensorflow/python/data/experimental/service/server_lib.py:206]: Thanks. Do you intend to add it to both the dispatcher and worker? It seems only added to the dispatcher, and the tests are failing with 

```
AttributeError: 'WorkerServer' object has no attribute 'stop'
```

**Review by yangustc07 — CHANGES_REQUESTED** [review]

File: tensorflow/python/data/experimental/service/server_lib.py (hunk 2 above)
Comment (EnricoMi) [reply tensorflow/python/data/experimental/service/server_lib.py:206]: Added `stop` to `WorkerServer`.

**Review by yangustc07 — APPROVED** [review]

**Review by yangustc07 — APPROVED** [review]

**Pull request merged**